bin/packages/watch.js

```diff
@@ -11,79 +11,154 @@ const chalk = require( 'chalk' );
+ */
+function isSourceFile( filename ) {
 	return (
+		/\/src\/.+\.(js|json|scss)$/.test( filename ) &&
 		! [
-			/\/(benchmark|__mocks__|__tests__|test|storybook|stories)\/.+.js$/,
+			/\/(benchmark|__mocks__|__tests__|test|storybook|stories)\/.+/,
 			/.\.(spec|test)\.js$/,
-		].some( ( regex ) => regex.test( filename ) ) &&
-		/.\.(js|json|scss)$/.test( filename )
+		].some( ( regex ) => regex.test( filename ) )
 	);
-};
+}
+
+/**
+ * Determine if a file is in a module package.
+ *
+ * getPackages only returns packages that have a package.json with the module
+ * field. Only build these packages.
+ *
+ * @param {string} filename
+ *
+ * @return {boolean} True if the file is in a module package.
+ */
+function isModulePackage( filename ) {
+	return modulePackages.some( ( packagePath ) => {
+		return filename.indexOf( packagePath ) > -1;
+	} );
+}
 
-const rebuild = ( filename ) => filesToBuild.set( filename, true );
+/**
+ * Is the file something the watch task should monitor or skip?
+ *
+ * @param {string} filename
+ * @param {symbol} skip
+ *
+ * @return {boolean | symbol} True if the file should be watched.
+ */
+function isWatchableFile( filename, skip ) {
+	return isSourceFile( filename ) && isModulePackage( filename )
+		? true
+		: skip;
+}
 
-getPackages().forEach( ( p ) => {
-	const srcDir = path.resolve( p, 'src' );
-	try {
-		fs.accessSync( srcDir, fs.F_OK );
-		watch(
-			path.resolve( p, 'src' ),
-			{ recursive: true, delay: 500 },
-			( event, filename ) => {
-				if ( ! isSourceFile( filename ) ) {
-					return;
-				}
+/**
+ * Returns the associated file in the build folder for a given source file.
+ *
+ * @param {string} srcFile
+ *
+ * @return {string} Path to the build file.
+ */
+function getBuildFile( srcFile ) {
+	// Could just use string.replace, but the user might have the project
+	// checked out and nested under another src folder.
+	const packageDir = srcFile.substr( 0, srcFile.lastIndexOf( '/src/' ) );
+	const filePath = srcFile.substr( srcFile.lastIndexOf( '/src/' ) + 5 );
+	return path.resolve( packageDir, 'build', filePath );
+}
+
+/**
+ * Adds a build file to the set of files that should be rebuilt.
+ *
+ * @param {'update'} event The event name
+ * @param {string} filename
+ */
+function updateBuildFile( event, filename ) {
+	if ( exists( filename ) ) {
+		try {
+			console.log( chalk.green( '->' ), `${ event }: ${ filename }` );
+			filesToBuild.set( filename, true );
+		} catch ( e ) {
+			console.log(
+				chalk.red( 'Error:' ),
+				`Unable to update file: ${ filename } - `,
+				e
+			);
+		}
+	}
+}
+
+/**
+ * Removes a build file from the build folder
+ * (usually triggered the associated source file was deleted)
+ *
+ * @param {'remove'} event The event name
+ * @param {string} filename
+ */
+function removeBuildFile( event, filename ) {
+	const buildFile = getBuildFile( filename );
+	if ( exists( buildFile ) ) {
+		try {
+			fs.unlinkSync( buildFile );
```

Comment:
Could probably use the async version of this method? I don't think it makes any difference in this context but could unblock the event loop when there're a lot of files being deleted at the same time (checking out a older branch for instance).

Reply:
Good suggestion 👍